### PR TITLE
Feature/49149: API endpoint for creating object

### PR DESCRIPTION
**Pull request by brunofarroporitense**
feature/49149_[BE] API endpoint for creating an Object:
- Adds version to ObjectManagement
-Changes liquidbase script to add Version in objectManagement.
-ObjectenApiPlugin now has a createObject function.
-Creates ObjectManagementProvider as a way to avoid circular dependencies.
-ZaakObjectResource now had a createObject post method.

### CODE REVIEW

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/ObjectenApiPlugin.kt [hunk 1]
```diff
@@ -61,4 +61,8 @@ class ObjectenApiPlugin(
     fun objectUpdate(objectUrl: URI, objectRequest: ObjectRequest): ObjectWrapper {
         return objectenApiClient.objectUpdate(authenticationPluginConfiguration, objectUrl, objectRequest)
     }
+
+    fun objectCreate(objectUrl: URI, objectRequest: ObjectRequest): ObjectWrapper {
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/ObjectenApiPlugin.kt:65]: I wonder why we started doing objectUpdate instead of updateObject, as the standard is the other way around.

Please change this to createObject, and I'll check on what to do with objectUpdate

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/client/ObjectenApiClient.kt [hunk 2]
```diff
@@ -96,4 +96,23 @@ class ObjectenApiClient(
 
         return result?.body!!
     }
+
+    fun objectCreate(
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/client/ObjectenApiClient.kt:100]: createObject

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/service/ZaakObjectService.kt [hunk 3]
```diff
@@ -134,8 +141,50 @@ class ZaakObjectService(
...
+                typeVersion = objectManagementProviderDTO.version,
+                data = data,
+                startAt = LocalDate.now()
+            )
+        )
+
+        return ZaakObjectDto(objectenApiPlugin.objectCreate(objectTypeUrl, createdObject).url)
+    }
+
+    // create
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/service/ZaakObjectService.kt:175]: Please remove these comments

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/service/ZaakObjectService.kt [hunk 4]
```diff
@@ -134,8 +141,50 @@ class ZaakObjectService(
         return zakenApiPluginInstance
     }
 
+    fun createObject(objectManagementId: UUID, data: JsonNode) : ZaakObjectDto {
+        val result = objectManagementProvider.getObjectManagementInfo(objectManagementId)
+
+        val objectManagementProviderDTO = ObjectManagementProviderDTO(
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/service/ZaakObjectService.kt:147]: This DTO is only used within this method, so it's not really transferring any data at all. Please use the values where needed, and skip the DTO stuff.

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/web/rest/ZaakObjectResource.kt [hunk 5]
```diff
@@ -63,4 +70,15 @@ class ZaakObjectResource(
         return form?.let { ResponseEntity.ok(it) } ?: ResponseEntity.notFound().build()
     }
 
+    @PostMapping(value = ["/v1/object"])
+    fun createZaakObjecten(
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/web/rest/ZaakObjectResource.kt:74]: Please use the singular 'createZaakObject'

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/service/ZaakObjectDto.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2015-2022 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/service/ZaakObjectDto.kt:2]: 2023

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/web/rest/ObjectManagementProvider.kt [hunk 7]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2015-2022 Ritense BV, the Netherlands.
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/web/rest/ObjectManagementProvider.kt:2]: 2023

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/client/ObjectenApiClient.kt [hunk 8]
```diff
@@ -77,7 +77,7 @@ class ObjectenApiClient(
         return result?.body!!
     }
 
-    fun objectUpdate(
+    fun updateObject(
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/client/ObjectenApiClient.kt:80]: Please revert the objectUpdate rename, as it will be a breaking change.

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/service/ZaakObjectService.kt (hunk 4 above)
Comment (theo-ritense) [reply zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/service/ZaakObjectService.kt:147]: Not sure why you marked it as resolved. The DTO is still there.

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/ObjectenApiPlugin.kt [hunk 9]
```diff
@@ -58,7 +58,11 @@ class ObjectenApiPlugin(
         )
     }
 
-    fun objectUpdate(objectUrl: URI, objectRequest: ObjectRequest): ObjectWrapper {
-        return objectenApiClient.objectUpdate(authenticationPluginConfiguration, objectUrl, objectRequest)
+    fun updateObject(objectUrl: URI, objectRequest: ObjectRequest): ObjectWrapper {
```
Comment (theo-ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/ObjectenApiPlugin.kt:61]: Please revert the change of objectUpdate, as it is a breaking change.

File: zgw/object-management/src/main/resources/config/liquibase/changelog/20221221-add-object-management-table.xml [hunk 10]
```diff
@@ -50,4 +50,13 @@
     <changeSet id="objecttype_id unique in object_management_configuration" author="Ritense">
         <addUniqueConstraint tableName="object_management_configuration" columnNames="objecttype_id" constraintName="uc_object_management_configuration_objecttype"/>
     </changeSet>
+
+    <changeSet id="add version column to object_management_configuration" author="Ritense">
+        <addColumn tableName="object_management_configuration">
```
Comment (marijnritense) [inline zgw/object-management/src/main/resources/config/liquibase/changelog/20221221-add-object-management-table.xml:55]: I think this was okay as object management isn't intended to be used yet... but just to be nice, maybe we can include a changeset that clears this table?

File: zgw/objecten-api/src/test/kotlin/com/ritense/objectenapi/service/ZaakObjectServiceTest.kt [hunk 11]
```diff
@@ -50,7 +51,8 @@ internal class ZaakObjectServiceTest {
     val zaakInstanceLinkService = mock<ZaakInstanceLinkService>()
     val pluginService = mock<PluginService>()
     val formDefinitionService = mock<FormDefinitionService>()
-    val zaakObjectService = ZaakObjectService(zaakInstanceLinkService, pluginService, formDefinitionService)
+    val objectManagementInfoProvider = mock<ObjectManagementInfoProvider>()
+    val zaakObjectService = ZaakObjectService(zaakInstanceLinkService, pluginService, formDefinitionService, objectManagementInfoProvider)
```
Comment (marijnritense) [inline zgw/objecten-api/src/test/kotlin/com/ritense/objectenapi/service/ZaakObjectServiceTest.kt:55]: This stuff would typically be done in an init method using the `BeforeEach` method. Do you mind moving the initialization there?

File: zgw/objecten-api/src/test/kotlin/com/ritense/objectenapi/web/rest/ZaakObjectResourceTest.kt [hunk 12]
```diff
@@ -181,6 +189,47 @@ internal class ZaakObjectResourceTest {
...
+
+        mockMvc.perform(
+            post("/api/v1/object?objectManagementId=$objectManagementId")
+                .characterEncoding(StandardCharsets.UTF_8.name())
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .accept(MediaType.APPLICATION_JSON_VALUE)
+                .content(data)
+        )
+            .andDo(MockMvcResultHandlers.print())
+            .andExpect(status().isCreated)
```
Comment (marijnritense) [inline zgw/objecten-api/src/test/kotlin/com/ritense/objectenapi/web/rest/ZaakObjectResourceTest.kt:211]: We also expect the URL to be set. Please add this as well.

File: zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/web/rest/ZaakObjectResource.kt [hunk 13]
```diff
@@ -63,4 +70,18 @@ class ZaakObjectResource(
         return form?.let { ResponseEntity.ok(it) } ?: ResponseEntity.notFound().build()
     }
 
+    @PostMapping(value = ["/v1/object"])
+    fun createZaakObject(
+        @RequestParam(name = "objectManagementId") objectManagementId: UUID,
+        @RequestBody data: JsonNode
+    ): ResponseEntity<Any> {
+        val objectDto = zaakObjectService.createObject(objectManagementId, data)
+        return objectDto.let {
```
Comment (Klaas-Ritense) [inline zgw/objecten-api/src/main/kotlin/com/ritense/objectenapi/web/rest/ZaakObjectResource.kt:79]:
```suggestion
        val objectUrl = zaakObjectService.createObject(objectManagementId, data)
        return objectUrl.let {
```

File: zgw/object-management/src/main/resources/config/liquibase/changelog/20221221-add-object-management-table.xml (hunk 10 above)
Comment (theo-ritense) [reply zgw/object-management/src/main/resources/config/liquibase/changelog/20221221-add-object-management-table.xml:55]: @ThomasRitense Do we need to clear the objectmanagement table, or not?

File: zgw/object-management/src/main/kotlin/com/ritense/objectmanagement/domain/ObjectManagement.kt [hunk 14]
```diff
@@ -27,26 +28,29 @@ import javax.persistence.Table
...
     @Column(name = "objecttypen_api_plugin_configuration_id", nullable = false)
-    val objecttypenApiPluginConfigurationId: UUID,
+    override val objecttypenApiPluginConfigurationId: UUID,
 
     @Column(name = "objecttype_id", nullable = false)
-    val objecttypeId: String,
+    override val objecttypeId: String,
+
+    @Column(name = "object_type_version")
+    override val objecttypeVersion: Int = 1,
```
Comment (Klaas-Ritense) [inline zgw/object-management/src/main/kotlin/com/ritense/objectmanagement/domain/ObjectManagement.kt:43]: The FE has a modal to create the ObjectManagement. That modal doesn't have the `objecttypeVersion` yet. 
Please either ask Kevin to update it, or update it yourself

File: zgw/object-management/src/main/resources/config/liquibase/changelog/20221221-add-object-management-table.xml (hunk 10 above)
Comment (Klaas-Ritense) [reply zgw/object-management/src/main/resources/config/liquibase/changelog/20221221-add-object-management-table.xml:55]: imo, the `defaultValue="1"` is good enough. Since that makes sure nothing breaks

Comment (ThomasRitense) [reply zgw/object-management/src/main/resources/config/liquibase/changelog/20221221-add-object-management-table.xml:55]: Agree with Klaas. This is not used in production yet and probably not even on test/acceptance.

File: zgw/object-management/src/main/kotlin/com/ritense/objectmanagement/domain/ObjectManagement.kt (hunk 14 above)
Comment (Klaas-Ritense) [reply zgw/object-management/src/main/kotlin/com/ritense/objectmanagement/domain/ObjectManagement.kt:43]: Or maybe a new story for this is needed

Comment (theo-ritense) [reply zgw/object-management/src/main/kotlin/com/ritense/objectmanagement/domain/ObjectManagement.kt:43]: Working on it.

**Review by Klaas-Ritense — APPROVED** [review]
